.design/requirements-575.md

```diff
@@ -0,0 +1,269 @@
+
+Implementations using the `SINGLETON` service scope must manage thread safety and session handling by themselves.
+The `BUNDLE` service scope should be considered as a `SINGLETON`.
+
+### UC2: Define and register a WebSocket server endpoint using the `Endpoint` class
+
+This is the equivalent of [UC1](#uc1-define-and-register-a-websocket-server-endpoint-using-annotations) but extending a class instead of using annotations.
+
+The Jakarta WebSocket specification provides the `Endpoint` class, which represents an object that can handle WebSocket conversations, both server-side or client-side.
+
+
+```java
+@Component(service = Endpoint.class, scope = ServiceScope.PROTOTYPE)
+@WebSocketWhiteboardServletPattern("/ws/work")
+public class WebSocketExtendsEndpointExample extends Endpoint {
+	public void onOpen(Session session, EndpointConfig config) {
+		// **Mandatory implementation**
+		// WebSocket stream is opened
+	}
+
+    public void onClose(Session session, CloseReason closeReason) {
+		// *Optional implementation*
+        // WebSocket stream is closed
+    }
+
+    public void onError(Session session, Throwable error) {
+		// *Optional implementation*
+        // Error handling
+    }
+}
+```
+
+The OSGi WebSocket Whiteboard implementation is responsible to register an instance of this service as WebSocket endpoint, either directly or using a proxy endpoint class.
+
+The message handler registration is done by the endpoint in the `onOpen` method.
+
+Like the handler of [UC1](#uc1-define-and-register-a-websocket-server-endpoint-using-annotations), it is expected that the endpoint service has a `PROTOTYPE` scope.
+
+#### Lack of URL parameters
+
+Extending the `Endpoint` class forbids to have extra parameters to the handling methods.
+As a result, it is not possible to have access to the URL parameters defined in the servlet pattern in that use case.
+
+### UC3: Define and register a WebSocket server message handler using an interface
+
+The Jakarta WebSocket specification provides the `MessageHandler.Whole<T>` interface to describe a message handler.
+It defines a single method `void onMessage(T message)` that will be called when the message has been fully received.
+
+The valid types for `T` are:
+* For text messages:
+  * `String`
+  * `Reader`
+  * `Decoder.Text`
+  * `Decoder.TextStream`
+* For binary messages:
+  * `ByteBuffer`
+  * `byte[]`
+  * `InputStream`
+  * `Decoder.Binary`
+  * `Decoder.BinaryStream`
+* For pong messages:
+  * `PongMessage`
+
+This interface lacks the `Session` argument, which is required to send data back to the peer.
+As a result, it might be preferred to define an OSGi specific interfaces that adds this argument and indicates the explicit kind of handled WebSocket frame.
+The OSGi WebSocket Whiteboard implementation will have to give the `Session` instance it obtains in the proxy message handler it registered.
```

Comment:
Message handlers are very restrictive on jakarta websocket spec, it seems not suitable to have any complex code in them. As per specification handlers are exspected to be added (only one handler per type!) to the session therefore I would not put Listeners in the responsibility of the Whiteboard Service.